diff --git a/bootstrap.conf b/bootstrap.conf
index a8ff2eef0..611ddf145 100644
--- a/bootstrap.conf
+++ b/bootstrap.conf
@@ -268,6 +268,7 @@ gnulib_modules="
   targetdir
   tempname
   termios-h
+  test-xfail
   time_rz
   timer-time
   timespec
diff --git a/tests/local.mk b/tests/local.mk
index 12e30b449..7504b3a14 100644
--- a/tests/local.mk
+++ b/tests/local.mk
@@ -751,6 +751,20 @@ all_tests =					\
   tests/touch/trailing-slash.sh			\
   $(all_root_tests)
 
+XFAIL_TESTS =
+if OS_IS_NETBSD
+XFAIL_TESTS += \
+  tests/rm/interactive-once.sh \
+  tests/shuf/shuf.sh \
+  tests/misc/sync.sh \
+  tests/tail/wait.sh \
+  tests/tail/pipe-f2.sh
+endif
+if OS_IS_SOLARIS11
+XFAIL_TESTS += \
+  tests/chgrp/from.sh
+endif
+
 # See tests/factor/create-test.sh.
 tf = tests/factor
 factor_tests = \
