diff --git a/src/sort.c b/src/sort.c
index 0928fd57c..997566240 100644
--- a/src/sort.c
+++ b/src/sort.c
@@ -2373,11 +2373,7 @@ debug_key (struct line const *line, struct keyfield const *key)
       if (key->sword != SIZE_MAX)
         beg = begfield (line, key);
       if (key->eword != SIZE_MAX)
-        {
-          lim = limfield (line, key);
-          /* Treat field ends before field starts as empty fields.  */
-          lim = MAX (beg, lim);
-        }
+        lim = limfield (line, key);
 
       if ((key->skipsblanks && key->sword == SIZE_MAX)
           || key->month || key_numeric (key))
